File: windows/Zikr/UpdateChecker.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Reflection;
using System.Web.Script.Serialization;

namespace Zikr
{
    /// <summary>
    /// Checks GitHub Releases for a newer tag than the running app's
    /// version. No in-place self-updater, same decision as the Linux
    /// build and for the same reason: installs vary (Inno Setup to
    /// Program Files, a portable extraction, a future winget/choco
    /// package), so silently overwriting files isn't a sound default.
    /// "Check for Updates" surfaces what changed and opens the releases
    /// page for you to download and reinstall.
    /// </summary>
    public static class UpdateChecker
    {
        private const string ReleasesUrl = "https://github.com/shamsbd71/zikr/releases/latest";
        private const string ApiUrl = "https://api.github.com/repos/shamsbd71/zikr/releases/latest";

        public enum Status { Available, UpToDate, Error }

        public sealed class CheckResult
        {
            public Status Status;
            public string Version;
            public string Message;
        }

        /// <summary>Reads the MSBuild `Version` property verbatim, as
        /// stamped into AssemblyInformationalVersionAttribute at build
        /// time via `-p:Version=X.Y.Z` (see release.yml/test.yml) -
        /// more direct than coercing through AssemblyVersion, which
        /// can't carry a value like "0.0.0-test".</summary>
        public static string CurrentVersion
        {
            get
            {
                var attrs = Assembly.GetExecutingAssembly()
                    .GetCustomAttributes(typeof(AssemblyInformationalVersionAttribute), false)
                    as AssemblyInformationalVersionAttribute[];
                return attrs != null && attrs.Length > 0 ? attrs[0].InformationalVersion : "0.0.0";
            }
        }

        public static CheckResult CheckForUpdate()
        {
            try
            {
                using (var client = new WebClient())
                {
                    client.Headers.Add("User-Agent", "Zikr-Updater");
                    client.Headers.Add("Accept", "application/vnd.github+json");
                    string json = client.DownloadString(ApiUrl);
                    var serializer = new JavaScriptSerializer();
                    var data = serializer.Deserialize<Dictionary<string, object>>(json);
                    string tag = data != null && data.TryGetValue("tag_name", out var t) ? Convert.ToString(t) : null;
                    if (string.IsNullOrEmpty(tag))
                        return new CheckResult { Status = Status.Error, Message = "Couldn't check for updates. Try again later." };

                    string latest = tag.StartsWith("v") ? tag.Substring(1) : tag;
                    string current = CurrentVersion;
                    if (CompareVersions(latest, current) > 0)
                        return new CheckResult { Status = Status.Available, Version = latest };
                    return new CheckResult { Status = Status.UpToDate, Version = current };
                }
            }
            catch (Exception)
            {
                return new CheckResult { Status = Status.Error, Message = "Couldn't check for updates. Try again later." };
            }
        }

        /// <summary>-1 if a&lt;b, 0 if equal, 1 if a&gt;b, comparing
        /// dot-separated numeric components.</summary>
        public static int CompareVersions(string a, string b)
        {
            int[] Parts(string v)
            {
                var segments = v.Split('.');
                var result = new int[segments.Length];
                for (int i = 0; i < segments.Length; i++)
                    int.TryParse(segments[i], out result[i]);
                return result;
            }

            var pa = Parts(a);
            var pb = Parts(b);
            int length = Math.Max(pa.Length, pb.Length);
            for (int i = 0; i < length; i++)
            {
                int va = i < pa.Length ? pa[i] : 0;
                int vb = i < pb.Length ? pb[i] : 0;
                if (va != vb) return va < vb ? -1 : 1;
            }
            return 0;
        }

        public static void OpenReleasesPage()
        {
            Process.Start(new ProcessStartInfo
            {
                FileName = ReleasesUrl,
                UseShellExecute = true,
            });
        }
    }
}

File: linux/zikr/update_checker.py
"""Checks GitHub Releases for a newer tag than the running version. No
in-place self-update, same decision as Windows: installs vary too much
across distros (.deb, a future AppImage/flatpak, a manual tarball) for
one safe overwrite strategy. "Check for Updates" surfaces what changed
and opens the release page for you to download and reinstall.
"""
import json
import urllib.request

API_URL = "https://api.github.com/repos/shamsbd71/zikr/releases/latest"
RELEASES_URL = "https://github.com/shamsbd71/zikr/releases/latest"


def compare_versions(a, b):
    """-1 if a<b, 0 if equal, 1 if a>b, comparing dot-separated numeric
    components (mirrors the macOS/Windows builds' numeric compare)."""
    def parts(v):
        return [int(p) if p.isdigit() else 0 for p in v.split(".")]

    pa, pb = parts(a), parts(b)
    n = max(len(pa), len(pb))
    pa += [0] * (n - len(pa))
    pb += [0] * (n - len(pb))
    if pa < pb:
        return -1
    if pa > pb:
        return 1
    return 0


def check_for_update(current_version, timeout=5):
    """Returns {"status": "available", "version": "1.5.0"} |
    {"status": "up_to_date"} | {"status": "error", "message": "..."}."""
    try:
        req = urllib.request.Request(API_URL, headers={"Accept": "application/vnd.github+json"})
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            data = json.loads(resp.read().decode("utf-8"))
    except Exception:
        return {"status": "error", "message": "Couldn't check for updates. Try again later."}

    tag = data.get("tag_name", "")
    latest = tag[1:] if tag.startswith("v") else tag
    if not latest:
        return {"status": "error", "message": "Couldn't check for updates. Try again later."}

    if compare_versions(latest, current_version) > 0:
        return {"status": "available", "version": latest}
    return {"status": "up_to_date"}
